include/clock_control.h

```diff
@@ -72,6 +90,67 @@ static inline int clock_control_off(struct device *dev,
 	return api->off(dev, sys);
 }
 
+/**
+ * @brief Request clock to start with notification when clock has been started.
+ *
+ * User can request delayed start by providing exact information when clock
+ * should be ready. Driver ensures that clock is ready before requested time.
+ * It is the driver responsibility to take into account clock startup time.
+ * When clock is already running user callback will be called from the context
+ * of the function call else it is called from other context (e.g. clock
+ * interrupt).
+ *
+ * @param dev 	  Device.
+ * @param sys	  A pointer to an opaque data representing the sub-system.
+ * @param data	  Data structure containing a callback that is called when
+ *		  action is performed. MUST NOT BE ON STACK!. Can be NULL.
+ * @param timeout When (given in k_cycle32_get()) when clock must be ready.
+ *		  Requested time must be within half of 32 bit span. Bigger
+ *		  values are considered as now request. Use k_cycle32_get for
+ *		  immediate request.
+ *
+ * @retval 0 if clock is already running.
+ * @retval -EALREADY if request has been scheduled.
+ * @retval -ENOTSUP if not supported.
+ */
+static inline int clock_control_async_on(struct device *dev,
+					 clock_control_subsys_t sys,
+					 struct clock_control_async_data *data,
+					 u32_t timeout)
+{
+	const struct clock_control_driver_api *api =
+		(const struct clock_control_driver_api *)dev->driver_api;
+
+	if (!api->async_on) {
+		return -ENOTSUP;
+	}
+
+	return api->async_on(dev, sys, data, timeout);
+}
+
+/**
+ * @brief Get clock status.
+ *
+ * @param dev Device.
+ *
+ * @retval 0 if clock is already running.
+ * @retval -EALREADY if request has been scheduled.
```

Comment:
I really preferred the enum in case get_status is implemented. -ENOTSUP is fine as an error, rest are just non erroneous status code.

Reply:
is it ok to return positive values as enum values and negative as error? I would prefer to keep function returning through value. It can be that enum will have status `unknown` which will mean that getting status is not supported.

Reply:
it is solved as described in https://github.com/zephyrproject-rtos/zephyr/pull/16675#discussion_r291968832